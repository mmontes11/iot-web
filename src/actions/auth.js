import {
  IS_AUTH,
  SET_USERNAME,
  SET_PASSWORD,
  LOGIN_REQUEST,
  LOGIN_REQUEST_SUCCESS,
  LOGIN_REQUEST_ERROR
} from "constants/actionTypes/auth";
import iotClient from "lib/iotClient";

export const isAuth = () => dispatch => {
  iotClient.authService
    .isAuth()
    .then(isAuthVal => dispatch({ type: IS_AUTH, isAuth: isAuthVal }))
    .catch(() => dispatch({ type: IS_AUTH, isAuth: false }));
};

export const setUsername = username => dispatch => {
  dispatch({ type: SET_USERNAME, username });
};

export const setPassword = password => dispatch => {
  dispatch({ type: SET_PASSWORD, password });
};

export const login = () => (dispatch, getState) => {
  dispatch({ type: LOGIN_REQUEST });
  const { username, password } = getState().auth;
  iotClient.authService.setCredentials(username, password);
  iotClient.authService
    .getToken()
    .then(res =>
      dispatch({ type: LOGIN_REQUEST_SUCCESS, statusCode: res.statusCode })
    )
    .catch(err =>
      dispatch({ type: LOGIN_REQUEST_ERROR, statusCode: err.statusCode, err })
    )
    .finally(() => isAuth()(dispatch));
};
