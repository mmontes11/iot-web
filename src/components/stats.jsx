import React, { Fragment } from "react";

const Stats = () => <Fragment>{null}</Fragment>;

export default Stats;
