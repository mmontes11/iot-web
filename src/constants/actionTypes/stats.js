export const STATS_UPDATED = "STATS_UPDATED";
export const STATS_REQUEST = "STATS_REQUEST";
export const STATS_REQUEST_SUCCESS = "STATS_REQUEST_SUCCESS";
export const STATS_REQUEST_ERROR = "STATS_REQUEST_ERROR";
