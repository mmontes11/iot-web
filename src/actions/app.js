import { TOGGLE_HAMBURGER_MENU, TOGGLE_MAP_DIALOG, SET_SHOW_ERROR } from "constants/actionTypes/app";

export const toggleHamburgerMenu = () => dispatch => {
  dispatch({ type: TOGGLE_HAMBURGER_MENU });
};

export const toggleMapDialog = () => dispatch => {
  dispatch({ type: TOGGLE_MAP_DIALOG });
};

export const setShowError = showError => dispatch => {
  dispatch({ type: SET_SHOW_ERROR, showError });
};
