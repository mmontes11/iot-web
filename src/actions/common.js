import { RESET } from "constants/actionTypes/common";

export const reset = () => dispatch => {
  dispatch({ type: RESET });
};
