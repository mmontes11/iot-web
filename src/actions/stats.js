import { STATS_UPDATED, STATS_REQUEST, STATS_REQUEST_SUCCESS, STATS_REQUEST_ERROR } from "constants/actionTypes/stats";
import { EVENT_TYPE, MEASUREMENT_TYPE } from "constants/observationTypes";
import { normalizeStats } from "helpers/statsNormalizer";
import iotClient from "lib/iotClient";

const requestStats = (type, observation, onStart, onSuccess, onError) => {
  if (type === EVENT_TYPE) {
    onStart();
    iotClient.eventService
      .getStatsByType(observation)
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  } else if (type === MEASUREMENT_TYPE) {
    onStart();
    iotClient.measurementService
      .getStatsByType(observation)
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  }
};

export const getStats = () => (dispatch, getState) => {
  const {
    stats: {
      params: {
        type: { selectedItem: selectedType },
        observation: { selectedItem: selectedObservation },
      },
    },
  } = getState();
  if (!selectedType || !selectedObservation) {
    return;
  }
  dispatch({ type: STATS_REQUEST });
  requestStats(
    selectedType,
    selectedObservation,
    () => dispatch({ type: STATS_REQUEST }),
    res => {
      dispatch({ type: STATS_REQUEST_SUCCESS });
      let stats;
      const responseStats = res.body.stats;
      if (responseStats) {
        stats = responseStats.map(normalizeStats);
      }
      dispatch({ type: STATS_UPDATED, stats });
    },
    () => dispatch({ type: STATS_REQUEST_ERROR }),
  );
};
