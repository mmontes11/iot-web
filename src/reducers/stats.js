import { RESET } from "constants/actionTypes/common";
import paramsReducer from "reducers/params";

const initialState = {
  stats: null,
};

export default (state = initialState, action) => {
  const params = paramsReducer(state.params, action);
  switch (action.type) {
    case RESET:
      return { ...initialState, params };
    default:
      return { ...state, params };
  }
};
