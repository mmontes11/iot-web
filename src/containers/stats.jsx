import React from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import { withResetOnUnmount } from "hocs/resetOnUnmount";
import { compose } from "recompose";
import QueryString from "query-string";
import PropTypes from "prop-types";
import StatsParamsPanel from "containers/statsParamsPanel";
import StatsFiltersPanel from "containers/statsFiltersPanel";
import BarChart from "components/barChart";
import Loader from "components/loader";
import * as statsActions from "actions/stats";
import * as commonActions from "actions/common";
import * as fromState from "reducers";
import { updateParams } from "actions/params";
import { addThingFilter } from "actions/thingFilter";
import { addTimePeriodFilter, addCustomTimePeriodFilter } from "actions/dateFilter";

const renderChart = (stats, isLoading) => {
  if (isLoading) {
    return <Loader />;
  } else if (stats) {
    return stats.map(statsItem => <BarChart key={statsItem.type} {...statsItem} />);
  }
  return null;
};

class Stats extends React.Component {
  componentDidMount() {
    const {
      match: {
        params: { type, observation },
      },
      location: { search },
    } = this.props;
    if (type && observation) {
      const queryParams = QueryString.parse(search);
      if (queryParams.thing) {
        this.props.addThingFilter(queryParams.thing);
      }
      if (queryParams.timePeriod) {
        this.props.addTimePeriodFilter(queryParams.timePeriod);
      } else if (queryParams.startDate || queryParams.endDate) {
        this.props.addCustomTimePeriodFilter(queryParams.startDate, queryParams.endDate);
      }
      this.props.updateParams(type, observation);
      this.props.getStats();
    }
  }
  componentDidUpdate() {
    const {
      match: {
        params: { type, observation },
      },
      hasError,
    } = this.props;
    if (type && observation && hasError) {
      this.props.history.push("/stats");
    }
  }
  _onParamsSelected = (type, observation) => {
    const { getStats, history, location } = this.props;
    history.push(`/stats/${type}/${observation}${location.search}`);
    getStats();
  };
  _onFiltersChange = (thing, timePeriod, startDate, endDate) => {
    const {
      getStats,
      location: { pathname },
    } = this.props;
    const params = {};
    if (thing) {
      params.thing = thing;
    }
    if (timePeriod) {
      params.timePeriod = timePeriod;
    }
    if (startDate) {
      params.startDate = startDate.toISOString();
    }
    if (endDate) {
      params.endDate = endDate.toISOString();
    }
    if (thing || timePeriod || startDate || endDate) {
      const search = Object.keys(params)
        .map(p => `${encodeURIComponent(p)}=${encodeURIComponent(params[p])}`)
        .join("&");
      this.props.history.push(`${pathname}?${search}`);
    } else {
      this.props.history.push(pathname);
    }
    getStats();
  };
  render() {
    const { stats, isLoading, reset } = this.props;
    return (
      <div className="container is-fluid section">
        <div className="columns">
          <div className="column is-three-quarters">
            <StatsParamsPanel onParamsSelected={this._onParamsSelected} onReset={() => reset()} />
            {renderChart(stats, isLoading)}
          </div>
          <div className="column is-one-quarter">
            <StatsFiltersPanel onFiltersChange={this._onFiltersChange} />
          </div>
        </div>
      </div>
    );
  }
}

Stats.propTypes = {
  stats: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  isLoading: PropTypes.bool.isRequired,
  hasError: PropTypes.bool.isRequired,
  getStats: PropTypes.func.isRequired,
  reset: PropTypes.func.isRequired,
  updateParams: PropTypes.func.isRequired,
  addThingFilter: PropTypes.func.isRequired,
  addTimePeriodFilter: PropTypes.func.isRequired,
  addCustomTimePeriodFilter: PropTypes.func.isRequired,
  match: PropTypes.shape({}).isRequired,
  location: PropTypes.shape({}).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
};

const withConnect = connect(
  state => ({
    stats: state.stats.items,
    isLoading: state.stats.isLoading,
    hasError: fromState.hasError(state),
  }),
  { ...statsActions, ...commonActions, updateParams, addThingFilter, addTimePeriodFilter, addCustomTimePeriodFilter },
);

export default compose(
  withConnect,
  withRouter,
  withResetOnUnmount,
)(Stats);
