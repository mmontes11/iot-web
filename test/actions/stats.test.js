import { getStats } from "actions/stats";
import { EVENT_TYPE, MEASUREMENT_TYPE } from "constants/observationTypes";
import { initialState } from "../constants";

describe("actions/stats", () => {
  it("dispatches a getStats action", () => {
    const thunk = getStats();
    const dispatch = jest.fn();
    let getState = jest.fn(() => ({
      ...initialState,
      stats: {
        ...initialState.stats,
        params: {
          type: { selectedItem: EVENT_TYPE },
          observation: { selectedItem: "door-opened" },
        },
      },
    }));
    thunk(dispatch, getState);
    expect(thunk).toBeAThunk();
    expect(thunk).toMatchSnapshot();
    getState = jest.fn(() => ({
      ...initialState,
      stats: {
        ...initialState.stats,
        params: {
          type: { selectedItem: MEASUREMENT_TYPE },
          observation: { selectedItem: "temperature-outdoor" },
        },
      },
    }));
    thunk(dispatch, getState);
    expect(thunk).toBeAThunk();
    expect(thunk).toMatchSnapshot();
  });
});
