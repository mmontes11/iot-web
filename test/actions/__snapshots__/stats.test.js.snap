// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions/stats dispatches a getStats action for event stats 1`] = `[Function]`;

exports[`actions/stats dispatches a getStats action for measurement stats 1`] = `[Function]`;
